frame/support/src/traits.rs

```diff
@@ -492,64 +492,91 @@ impl<
+/// the best estimate possible these methods should be called throughout the duration of the session
+/// (rather than calling once and storing the result).
+///
 /// This should be the same logical unit that dictates `ShouldEndSession` to the session module. No
-/// Assumptions are made about the scheduling of the sessions.
+/// assumptions are made about the scheduling of the sessions.
 pub trait EstimateNextSessionRotation<BlockNumber> {
 	/// Return the average length of a session.
 	///
 	/// This may or may not be accurate.
 	fn average_session_length() -> BlockNumber;
 
+	/// Return an estimate of the current session progress.
+	///
+	/// None should be returned if the estimation fails to come to an answer.
+	fn estimate_current_session_progress(now: BlockNumber) -> Option<Percent>;
+
 	/// Return the block number at which the next session rotation is estimated to happen.
 	///
-	/// None should be returned if the estimation fails to come to an answer
+	/// None should be returned if the estimation fails to come to an answer.
 	fn estimate_next_session_rotation(now: BlockNumber) -> Option<BlockNumber>;
 
-	/// Return the weight of calling `estimate_next_session_rotation`
-	fn weight(now: BlockNumber) -> Weight;
+	/// Return the weight of calling `estimate_current_session_progress`.
+	fn estimate_current_session_progress_weight(now: BlockNumber) -> Weight;
+
+	/// Return the weight of calling `estimate_next_session_rotation`.
+	fn estimate_next_session_rotation_weight(now: BlockNumber) -> Weight;
 }
 
-impl<BlockNumber: Bounded + Default> EstimateNextSessionRotation<BlockNumber> for () {
+impl<BlockNumber: Zero> EstimateNextSessionRotation<BlockNumber> for () {
 	fn average_session_length() -> BlockNumber {
-		Default::default()
+		Zero::zero()
+	}
+
+	fn estimate_current_session_progress(_: BlockNumber) -> Option<Percent> {
+		None
 	}
 
 	fn estimate_next_session_rotation(_: BlockNumber) -> Option<BlockNumber> {
-		Default::default()
+		None
 	}
 
-	fn weight(_: BlockNumber) -> Weight {
-		0
+	fn estimate_current_session_progress_weight(_: BlockNumber) -> Weight {
+		Zero::zero()
+	}
+
+	fn estimate_next_session_rotation_weight(_: BlockNumber) -> Weight {
+		Zero::zero()
 	}
 }
 
-/// Something that can estimate at which block the next `new_session` will be triggered.
+/// Something that can estimate at which block the next `new_session` will be triggered, i.e. when
+/// we will try to fetch new validators. For example, if we are using a staking module this would be
+/// the block when the session module would ask staking what the next validator set will be, as such
+/// this must always be implemented by the session module.
 ///
-/// This must always be implemented by the session module.
+/// With `pallet-session` fetching the new validators happens when we rotate sessions but different
+/// implementations of session could have different logic, still any estimate coming from here should
```

Comment:
Please validate that these comments are indeed correct :)

Reply:
> still any estimate coming from here should always be lower or equal to the estimates for next session rotates.

I'm not sure I understand this.

I think the whole confusion come from names, here a better name would be EstimateNextScheduleSession, average_session_length should be renamed average_blocks_between_scheduled_session.

If the pallet session would trigger election of validator set 1 session and half before using it for validating (so half a session before setting it as queued validator in `SessionHandler::on_new_session`), then `estimate_next_new_session` can be more than `estimate_next_session_rotation`.

Maybe those trait should merge finally, but what I wanted is that:
* staking needs to know when the next election will happen (so when SessionManager::new_session is called)
* only pallet-session can say when SessionManager::new_session is called so better to have session implement the the trait.

But I know that we plan a new session at the same time that we rotate sessions, so as long as this hold, we could just give Babe implementation of EstimateNextSessionRotation to pallet-staking.

Note that there is an issue for some renamings, https://github.com/paritytech/substrate/issues/7742

Apart from the last line, its all good

Reply:
I reformulated the doc and removed the last sentence. I kept the name of the trait as is and we should rename as part of #7742 later on. The way I am thinking `average_blocks_between_scheduled_session` should always be equal to `average_session_length`, they can be offset from one another (for example we fetch validators one session and a half in advance), but the period should be the same right?

Reply:
Yes I think so